### PR TITLE
MNIST example based on arn_generic.py

### PR DESCRIPTION
**Pull request by akrentz6**
An example, based on `examples/arn_generic.py`, to use `xlns` objects when training a neural network. This rewrite includes bug fixes and cleans up the code. The biggest changes are:
- Gets rid of the long if statements that cast to different objects based on the `type`. I use a dictionary with functions to cast to the correct type.
- Before, if the batchsize didn't divide the number of samples, the final, smaller batch would be ignored but I add support to include this.
- `examples/arn_generic.py` has an inconsistent use of tabs and spaces which causes errors, this script doesn't.
- Before, there was tons of unnecessary type casting, e.g. from `ndarray` to `list` back to `ndarray`, etc. I got rid of this.
- There were a lot of unnecessary variables being defined. For example, `delta_W1` and `delta_W2` are defined then never used and further, `lnsdelta_W1` and `lnsdelta_W2` are defined but then rewritten over during training.
- I added a bunch more arguments (all the defaults are kept the same):
  - `--mnist`, `--weight_in`, `--weight_out`, `--figure_path`, so users can specify paths to files.
  - `--weight_scale`, the standard deviation for weights if they're generated.
  - `--figure`, for if users want a figure to be generated or not.
  - `--precision`, `--weights_precision`, precision for xlns objects.
- Since before, the script would do nothing if `--is_training` evaluated to false, now we test the model on the validation set.

### CODE REVIEW

**Comment by akrentz6** [comment]
Not sure why the checks failed, the warnings that the checks show are to do with other files but this PR only adds a file to examples, it doesn't touch anything else. Maybe you guys will know what's happened.

**Comment by markgarnold** [comment]
@pradeeban This looks like a nice addition to examples, but I don't understand the build fail.  I am reluctant to merge until we get to the bottom of that.

**Comment by pradeeban** [comment]
Lint with flake8 says:
2s
Run # stop the build if there are Python syntax errors or undefined names
./examples/arnnpr_lpvip.py:31:3: F824 `global xlnsF` is unused: name is never assigned in scope
  global xlnsF
  ^
./src/xlns.py:23:3: F824 `global xlnsF` is unused: name is never assigned in scope
  global xlnsB,xlnsF
  ^
./src/xlns.py:48:3: F824 `global xlnsB` is unused: name is never assigned in scope
  global xlnsB
  ^
./src/xlns.py:57:3: F824 `global xlnsB` is unused: name is never assigned in scope
  global xlnsB
  ^
./src/xlns.py:258:3: F824 `global xlnsB` is unused: name is never assigned in scope
  global xlnsB
  ^
./src/xlns.py:327:3: F824 `global xlnsB` is unused: name is never assigned in scope
  global xlnsB
  ^
./src/xlns.py:737:3: F824 `global ovfmin` is unused: name is never assigned in scope
  global ovfmin,ovfmax
  ^
./src/xlns.py:737:3: F824 `global ovfmax` is unused: name is never assigned in scope
  global ovfmin,ovfmax
  ^
./src/xlns.py:2075:3: F824 `global ovfmin` is unused: name is never assigned in scope
  global ovfmin,ovfmax
  ^
./src/xlns.py:2075:3: F824 `global ovfmax` is unused: name is never assigned in scope
  global ovfmin,ovfmax
  ^
./src/xlns.py:2629:3: F824 `global ovfmin` is unused: name is never assigned in scope
  global ovfmin,ovfmax
  ^
./src/xlns.py:2629:3: F824 `global ovfmax` is unused: name is never assigned in scope
  global ovfmin,ovfmax
  ^
12    F824 `global xlnsF` is unused: name is never assigned in scope
12
Error: Process completed with exit code 1.

**Comment by markgarnold** [comment]
OK.  I am wondering if this has something to do with the fact that the process that builds the wheel starts with xlns and xlnsconf in /src, but a later `pip install xlns` then allows `import xlns as xl` and `import xlnsconf.whatever` legal from the "top level".

It is strange this does not trip up arn_generic.py.   I had thought we had exempted examples from being tested to allow this.

This is probably a simple pypi/github problem, but I am not sure where to look.  Anyone have a thought?
